수강신청 미션 - 1단계 리뷰 요청합니다!

jonghunbaek opened the pull request:
안녕하세요 규남님! 마지막 미션을 하게 된 백종훈입니다.  이번 미션 동안 잘 부탁 드리겠습니다😃  

이번 미션을 진행하며 궁금한 점이 생겨 질문 남깁니다!
1. 일급 컬렉션을 속성으로 가지고 있는 객체의 경우 getter로 일급 컬렉션을 꺼내오는 것도 좋지 못한 방법인가요? 일급 컬렉션 자체가 포장되어 있기 때문에 getter로 꺼내어도 괜찮지 않나 의문이 들어서 여쭤봅니다!

2. Answers deleteAll()에서 answers를 foreach()로 answer의 상태를 변경하는 것은 이펙티브 자바 46에 따르면 안티패턴이라고 생각되는데 다른 방법이 있을까요? 이런 경우엔 일반적으로 for문으로 해결을 하나요?

3. Question의 테스트 코드중 deleteHistories에서 실제와 유사하게 테스트를 진행하려다 보니 given절이 커졌습니다. Answers의 삭제 기록 생성에 대한 테스트는 AnswersTest에서 진행하니 해당 부분을 의도적으로 빼도 괜찮은 건가요? 빼지 않는다면 Mock객체를 활용해 코드를 간소화하는 것이 최선일까요?

이상입니다. 이번 리뷰 잘 부탁 드립니다👍

jonghunbaek committed:
docs : readme 요구사항 분석 업데이트

jonghunbaek committed:
feat : 사용자 정보를 인자로 받아 답변의 작성자와 일치하는 지 확인한다.

jonghunbaek committed:
feat : 질문을 삭제할 때 사용자 정보를 인자로 받아 검증 조건을 위반하면 예외를 던진다.

jonghunbaek committed:
feat : Answer의 deleted속성을 true로 바꿔 답변을 삭제한다.

jonghunbaek committed:
feat : 조건문 메서드 추출 및 답변 삭제

jonghunbaek committed:
feat : Answer 일급 컬렉션 Answers 생성 및 Question 필드 컴파일 오류 수정

jonghunbaek committed:
feat : DeleteHistory를 만들어 반환한다.

jonghunbaek committed:
feat : 삭제되지 않은 답변의 삭제기록을 생성하면 예외를 발생시킨다.

jonghunbaek committed:
feat : DeleteHistory 리스트를 만들어 반환한다.

jonghunbaek committed:
feat : DeleteHistory 리스트를 만들어 반환한다.

- Answers 초기화 문제 수정

jonghunbaek committed:
feat : 불필요 메서드 제거 및 테스트 코드 오류 수정

jonghunbaek committed:
docs : readme 업데이트

jonghunbaek committed:
test : 예외에 대한 누락 테스트 추가

kyucumber commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Answers {
```
> 일급 컬렉션을 속성으로 가지고 있는 객체의 경우 getter로 일급 컬렉션을 꺼내오는 것도 좋지 못한 방법인가요? 일급 컬렉션 자체가 포장되어 있기 때문에 getter로 꺼내어도 괜찮지 않나 의문이 들어서 여쭤봅니다!

getter를 지양하고 메서드를 이용해 명확한 메시지를 표현하는게 좋지만 단순히 데이터를 가져가야 하는 경우도 얼마든지 있을 수 있어서 필요에 따라 getter를 이름 그대로 구현할 수 있습니다. 다만 참조가 노출되어 외부에서 수정하거나 하는 상황에 대해서는 Collections.unmodifiableList 혹은 Stream collect를 활용해 반환해서 막아주셔야 할 것 같습니다.

kyucumber commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,51 @@
+        this.answers = new ArrayList<>(answers);
+    }
+
+    public void deleteAll(NsUser loginUser) throws CannotDeleteException {
```
> Answers deleteAll()에서 answers를 foreach()로 answer의 상태를 변경하는 것은 이펙티브 자바 46에 따르면 안티패턴이라고 생각되는데 다른 방법이 있을까요? 이런 경우엔 일반적으로 for문으로 해결을 하나요?

아이템 46의 내용이 스트림에서 부작용 없는 함수를 사용하라는 내용이었던걸로 기억합니다.

사실 일반적으로 여기저기서 참조를 가져다 쓸 수 있는 컬렉션이라면 당연히 종단 연산이 아닌 forEach 내부에서 delete를 수행하는것이 문제가 되겠지만 일급 컬렉션 내부에서는 해당 참조를 숨기고 있기 때문에 큰 문제가 될 것이라고는 생각되지 않습니다.

굳이 for 문을 사용할 필요가 없어보입니다. for문이라고 문제가 없는것도 아니라서요.

kyucumber commented on src/test/java/nextstep/qna/domain/QuestionTest.java:
```diff
@@ -1,8 +1,92 @@
+            .hasMessage("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
+    }
+
+    @DisplayName("현재 시간을 인자로 받아 질문, 답변의 삭제 기록 리스트를 반환한다.")
```
> Question의 테스트 코드중 deleteHistories에서 실제와 유사하게 테스트를 진행하려다 보니 given절이 커졌습니다. Answers의 삭제 기록 생성에 대한 테스트는 AnswersTest에서 진행하니 해당 부분을 의도적으로 빼도 괜찮은 건가요? 빼지 않는다면 Mock객체를 활용해 코드를 간소화하는 것이 최선일까요?

다른 쪽에서 명확히 테스트가 된 부분이라면 빼도 괜찮습니다. 다만 AnswerTest에 포함된 경우라 어떤 변경 등이 누락될 수 있을 것 같아 이런 경우라면 DeleteHistory 테스트를 별도로 만들거나 하는게 더 나을수도 있겠네요.

그리고 테스트를 위해 필요한 데이터를 준비하는 것은 정해진 답이 없습니다. 결국 다 만들어야 하는건 똑같아서 말씀하신 것 처럼 Mock으로 테스트할 부분을 명확히 분리하거나 아니면 직접 모든 객체에 대한 Fixture를 만드는 것 밖에 없습니다.

kyucumber commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -47,31 +49,34 @@ public Long getId() {
-        return contents;
-    }
-
     public void toQuestion(Question question) {
```
toQuestion이라고 하니 Answer을 Question으로 변환해 반환하는 것 처럼 보이기도 하네요.

Answer 내부의 Question을 할당하는 것이므로 차라리 명확하게 setQuestion이 나아보입니다. setter나 getter를 지양하라고 하지만 단순히 getter, setter를 남발하고 메서드를 추출하지 않는 경우에 대해 지양하라는 것이기 떄문에 필요한 경우엔 사용해도 무방하다고 생각해요.

kyucumber commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -47,31 +49,34 @@ public Long getId() {
     }
 
+    public boolean isSameWriter(NsUser loginUser) {
+        return writer.equals(loginUser);
```
NsUser에는 equals 및 hashCode가 제대로 구현되어 있나요~?

kyucumber commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -47,31 +49,34 @@ public Long getId() {
+        this.deleted = true;
+    }
+
+    public DeleteHistory deleteHistory(LocalDateTime now) {
```
일반적인 메서드 네이밍 패턴을 고려하면 History를 삭제하는 것으로 오인되기 쉬워보입니다. 차라리 명확하게 getDeleteHistory 등의 네이밍을 가져가거나 혹은 더 적절한 이름을 고민해보시면 좋겠습니다.

kyucumber replied:
삭제 이력은 삭제 시점에 정해져야 할 것 같은데요. 외부에서 override 하지 않고 내부에서 LocalDateTime.now()를 수행하는 것은 어떻게 생각하시나요?

kyucumber commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -47,31 +49,34 @@ public Long getId() {
+            return new DeleteHistory(ANSWER, id, writer, now);
+        }
+
+        throw new IllegalArgumentException("해당 답변은 삭제되지 않았습니다.");
```
프로덕션 환경에서 이런 예외가 발생하면 어떤 답변인지 찾기가 어렵습니다. answer의 id 등을 남겨 추적하기 쉽도록 하면 좋을 것 같네요.

kyucumber commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,51 @@
+
+    private void validateOtherWriter(NsUser loginUser) throws CannotDeleteException {
+        if (isOtherUser(loginUser)) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
```
이 예외는 CheckedException으로 정의되어 있습니다. 혹시 어떤 의도로 이를 Checked Exception으로 정의했는지 고민해보시고 사용해보셨을까요~?

kyucumber commented on src/test/java/nextstep/qna/domain/AnswerTest.java:
```diff
@@ -1,8 +1,64 @@
+    @DisplayName("사용자 정보를 인자로 받아 답변 작성자와 일치하는 지 확인한다.")
+    @Test
+    void isSameWriter() {
+        // given
```
테스트 블럭을 깔끔하게 잘 나누어주셨네요 👍

kyucumber commented on src/test/java/nextstep/qna/domain/AnswerTest.java:
```diff
@@ -1,8 +1,64 @@
 public class AnswerTest {
-    public static final Answer A1 = new Answer(NsUserTest.JAVAJIGI, QuestionTest.Q1, "Answers Contents1");
-    public static final Answer A2 = new Answer(NsUserTest.SANJIGI, QuestionTest.Q1, "Answers Contents2");
+    public static final Answer A1 = new Answer(JAVAJIGI, QuestionTest.Q1, "Answers Contents1");
+    public static final Answer A2 = new Answer(SANJIGI, QuestionTest.Q1, "Answers Contents2");
```
이런 경우 이 테스트 클래스를 참조하게 될텐데 테스트 컨텍스트가 오염될 수 있을 것 같아요.
Fixture를 별도로 추출해서 다른 곳에서도 활용될 수 있도록 하면 좋을 것 같습니다.

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

jonghunbaek committed:
feat : equals,hashCode 오버라이딩

jonghunbaek committed:
rename : 변수명 변경 deleteHistory -> createDeleteHistory

jonghunbaek committed:
refactor : test fixture 별도 클래스로 분리

jonghunbaek committed:
refactor : toQuestion() 삭제 및 addAnswer접근 제어자 protected로 변경

- Answer생성 시 Question의 상태도 한번에 변경하도록 유도
- addAnswer()는 다른 레이어에선 직접 추가하지 못하도록 protected로 설정

jonghunbaek committed:
refactor : delete()호출 시 인자로 현재 시간을 넘겨 삭제 시간을 저장, CannotDeleteException을 unchecked 예외로 변경

- 삭제 이력 생성 시 현재 시간을 넘겨주는 것은 실제 삭제 시간이랑 일치하지 않을 수 있기 때문
- CannotDeleteException의 경우 예외처리를 강제할 필요가 없기 때문에 unchecked로 수정

jonghunbaek committed:
fix : warning제거

jonghunbaek committed:
refactor : UnChecked -> Checked로 재변경

jonghunbaek commented:
안녕하세요! 1단계 리뷰 재요청합니다!
규남님의 저번 피드백을 통해 기존 코드가 '옳다'는 생각으로 리팩토링 하는 태도를 놓게 됐습니다. 덕분에 다양한 고민을 해볼 수 있게 돼서 감사합니다😄  

> 이 예외는 CheckedException으로 정의되어 있습니다. 혹시 어떤 의도로 이를 Checked Exception으로 정의했는지 고민해보시고 사용해보셨을까요~?

앞서 말한 것처럼 기존 코드를 그대로 적용하게 됐습니다. 하지만 피드백을 보고 고민한 결과 Checked가 타당하다고 판단했습니다.    판단 기준은 아래와 같습니다. 1번의 기준 때문에 UnChecked를 고민 했지만 3번의 기준이 판단에 결정적인 기준이 됐습니다. 이외에 또 알아두면 좋을 기준이 있을까요?
``` 
Checked vs UnChecked
1. 컴파일 타임 vs 런타임 
2. 예외 처리 강제 vs 강제x
3. 사용자 입력 같은 외적 요인 vs 프로그래머의 실수와 같은 내부적 요인
```

> 이런 경우 이 테스트 클래스를 참조하게 될텐데 테스트 컨텍스트가 오염될 수 있을 것 같아요.
Fixture를 별도로 추출해서 다른 곳에서도 활용될 수 있도록 하면 좋을 것 같습니다.

먼저, 해당 피드백의 의도를 Fixture를 위한 클래스 별도 생성이라고 받아 드렸는데 오해 없이 받아 드린 걸까요?
더불어 평소 static한 Fixture의 사용을 지양하는 편입니다. 편리하긴 하지만 종종 의도와 다르게 테스트의 정합성이 깨지고 이를 수정하는 시간이 꽤 오래 걸린 적이 있기 때문입니다. 
이와 관련해서 실제 팀 단위로 코드를 작성할 때 Fixture의 사용을 전역으로 설정하는 경우가 일반적인가요? 만약 전역으로 사용하게 된다면 관련한 문제들이 팀 내에서 종종 발생할 것으로 예상되는데 주로 어떤 해결책을 사용하시나요?  

바쁘신 와중에도 꼼꼼한 피드백 감사합니다. 오늘도 좋은 하루 보내십쇼👍

jonghunbaek committed:
refactor : 답변 삭제 시 검증을 일급 컬렉션 -> Answer로 이동

- 관련 테스트 수정 및 불필요 테스트 삭제

jonghunbaek commented:
어제 라이브 강의 듣고 추가로 수정할 부분이 생겨 추가 commit했습니다!!

> 삭제 이력은 삭제 시점에 정해져야 할 것 같은데요. 외부에서 override 하지 않고 내부에서 LocalDateTime.now()를 수행하는 것은 어떻게 생각하시나요?

그리고 이 부분에 대한 답변을 누락해 같이 추가로 남깁니다.
1. QnAServiceTest에서 verify로 현재 시간을 검증하지만 mock객체이기 때문에 도메인에서 별도의 테스트를 한 번 더 해줘야 한다고 판단했습니다
2. delete()와 createDeleteHistories()는 다른 기능이기에 두 개의 메서드로 분리했습니다.
3. 메서드를 2개로 분리하니 삭제 시점, 삭제 기록 생성 시점의 차이가 생기고 피드백을 반영해 삭제 시점에 현재 시간을 인자로 받았습니다.
4. 다만 이 방식엔 Answer, Question에 deletedDate가 삭제 전까지는 초기화가 안되는 단점이 존재합니다. 
5. 4에서 언급한 단점보단 테스트의 안정성이 중요하다고 판단해 해당 방식을 택했습니다.
6. 하지만 어느 방식이 더 치명적인 단점일 지는 명확하게 판단하기 어려웠습니다. 이 부분에 대해 규남님의 추가적인 의견을 듣고 싶습니다!!

kyucumber commented:
답변을 남기던 도중 누락되었네요. 기존에 잘못 남긴 건 삭제하고 다시 답변 남겨둘게요.

> 앞서 말한 것처럼 기존 코드를 그대로 적용하게 됐습니다. 하지만 피드백을 보고 고민한 결과 Checked가 타당하다고 판단했습니다. 판단 기준은 아래와 같습니다. 1번의 기준 때문에 UnChecked를 고민 했지만 3번의 기준이 판단에 결정적인 기준이 됐습니다. 이외에 또 알아두면 좋을 기준이 있을까요?

말씀하신 기준 정도면 충분할 것 같습니다.

> 먼저, 해당 피드백의 의도를 Fixture를 위한 클래스 별도 생성이라고 받아 드렸는데 오해 없이 받아 드린 걸까요?
더불어 평소 static한 Fixture의 사용을 지양하는 편입니다. 편리하긴 하지만 종종 의도와 다르게 테스트의 정합성이 깨지고 이를 수정하는 시간이 꽤 오래 걸린 적이 있기 때문입니다.
> 이와 관련해서 실제 팀 단위로 코드를 작성할 때 Fixture의 사용을 전역으로 설정하는 경우가 일반적인가요? 만약 전역으로 사용하게 된다면 관련한 문제들이 팀 내에서 종종 발생할 것으로 예상되는데 주로 어떤 해결책을 사용하시나요?

일단 저는 팀 내에서 코틀린을 사용하고 있으며 이에 따라 data class의 copy를 활용할 수 있습니다. copy를 수행하게 되면 static이어도 서로 영향을 미치지 않아 이런 이슈가 많진 않은 것 같아요. 
그리고 그와 별개로 저는 object로 create 메서드를 통해 기본 값을 가진 Fixture를 만드는 형태로 사용합니다. static 변수를 직접 바라보더라도 사실 테스트 컨텍스트의 라이프사이클을 잘 가져가면 크게 문제는 발생하지 않을 것 같은데 가급적이면 메서드를 통해 Fixture를 그 시점에 생성할 수 있도록 구현하지 static 변수를 가져다 쓰는 것을 지양하는 편입니다.

kyucumber commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -40,36 +45,41 @@ public Answer(Long id, NsUser writer, Question question, String contents) {
+        }
+    }
+
+    public DeleteHistory createDeleteHistory() {
```
> QnAServiceTest에서 verify로 현재 시간을 검증하지만 mock객체이기 때문에 도메인에서 별도의 테스트를 한 번 더 해줘야 한다고 판단했습니다 delete()와 createDeleteHistories()는 다른 기능이기에 두 개의 메서드로 분리했습니다. 메서드를 2개로 분리하니 삭제 시점, 삭제 기록 생성 시점의 차이가 생기고 피드백을 반영해 삭제 시점에 현재 시간을 인자로 받았습니다.
다만 이 방식엔 Answer, Question에 deletedDate가 삭제 전까지는 초기화가 안되는 단점이 존재합니다.
> 4에서 언급한 단점보단 테스트의 안정성이 중요하다고 판단해 해당 방식을 택했습니다. 하지만 어느 방식이 더 치명적인 단점일 지는 명확하게 판단하기 어려웠습니다. 이 부분에 대해 규남님의 추가적인 의견을 듣고 싶습니다!!

이 부분은 잘 이해가 안되는데 추가 설명을 해주실 수 있을까요? deletedDate가 삭제 전에 초기화가 되어야 할 이유를 잘 모르겠습니다. 삭제하는 시점에 초기화가 되는게 정상적인 흐름으로 보여서요.

null과 같은 값을 가지고 있는게 문제라고 생각하시는거라면 문제가 없을 것 같습니다. JPA 등을 사용해 엔티티의 삭제 시간 등을 기록하더라도 nullable하게 가져가는 경우가 많아서요. 원하시는 답변이 아니라면 별도로 다음 단계 혹은 DM으로 문의 주시면 감사하겠습니다.

jonghunbaek replied:
> 이 부분은 잘 이해가 안되는데 추가 설명을 해주실 수 있을까요? deletedDate가 삭제 전에 초기화가 되어야 할 이유를 잘 모르겠습니다. 삭제하는 시점에 초기화가 되는게 정상적인 흐름으로 보여서요.

아래 문단에서 설명해주시는 것처럼 null값을 가지고 있는 것이 문제라고 생각했었기 때문입니다! 답변 덕분에 하나 더 알게 돼서 감사드립니다😄

kyucumber reviewed (approved):
수고 많으셨습니다.

질문에 답변 남겨두었으니 다음 단계 진행에 참고 부탁드려요.
